src/main/java/org/chameleoncloud/GlobusUserAttributeMapper.java

```diff
@@ -0,0 +1,109 @@
+
+package org.chameleoncloud;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+
+import org.chameleoncloud.representations.GlobusIdentity;
+import org.jboss.logging.Logger;
+import org.keycloak.broker.oidc.OIDCIdentityProviderFactory;
+import org.keycloak.broker.oidc.mappers.AbstractClaimMapper;
+import org.keycloak.broker.provider.BrokeredIdentityContext;
+import org.keycloak.models.IdentityProviderMapperModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserModel;
+import org.keycloak.provider.ProviderConfigProperty;
+import org.keycloak.util.JsonSerialization;
+
+public class GlobusUserAttributeMapper extends AbstractClaimMapper {
+    public static final String PROVIDER_ID = "globus-user-attribute-mapper";
+    private static final String[] cp = new String[] { OIDCIdentityProviderFactory.PROVIDER_ID };
+    private static final List<ProviderConfigProperty> configProperties = new ArrayList<>();
+    private static final Logger logger = Logger.getLogger(GlobusUserAttributeMapper.class);
+
+    public static final String BROKER_ATTRIBUTE = "broker.attribute";
+    public static final String USER_ATTRIBUTE = "user.attribute";
+    public static final String IDENTITY_SET = "identity_set";
+    public static final String SUB_LINKED = "linked";
+
+    public static final String getSubAttribute(String providerId, String sub) {
+        return String.join("_", providerId, "sub", sub);
+    }
+
+    public Set<String> getLinkedSubs(String providerId, UserModel user) {
+        // Get all attributes from user
+        String prefix = String.join("_", providerId, "sub");
+        Map<String, List<String>> attributes = user.getAttributes();
+        Set<String> linkedSubs = attributes.keySet().stream().filter(k -> k.startsWith(prefix))
+                .map(k -> user.getFirstAttribute(k)).collect(Collectors.toSet());
+
+        return linkedSubs;
+    }
+
+    private void addLinkedSubs(String providerId, UserModel user, Set<String> tokenSubs) {
+        for (String sub : tokenSubs) {
+            // Set Key in the form <alias>_sub_<sub>: linked
+            // We don't store list due to max length of DB field
+            user.setSingleAttribute(getSubAttribute(providerId, sub), SUB_LINKED);
```

Comment:
Instead of just storing "linked", would it make sense to store the preferred username for that identity?

Reply:
I'm using the key+value for the "search for user by attribute" method.
While I think? that the username should always match, I'm not entirely sure of the implications.